src/Storages/TimeSeries/PrometheusHTTPProtocolAPI.cpp

```diff
@@ -0,0 +1,500 @@
+#include <Storages/TimeSeries/PrometheusHTTPProtocolAPI.h>
+
+#include <Common/logger_useful.h>
+#include <Common/thread_local_rng.h>
+#include <Core/Field.h>
+#include <IO/WriteBufferFromString.h>
+#include <IO/WriteHelpers.h>
+#include <Storages/StorageTimeSeries.h>
+#include <Parsers/Prometheus/PrometheusQueryTree.h>
+#include <Parsers/Prometheus/PrometheusQueryResultType.h>
+#include <Storages/TimeSeries/PrometheusQueryToSQL.h>
+#include <Storages/TimeSeries/TimeSeriesColumnNames.h>
+#include <Interpreters/executeQuery.h>
+#include <Interpreters/Context.h>
+#include <Core/Settings.h>
+#include <Processors/Executors/PullingPipelineExecutor.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeTuple.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/DataTypesDecimal.h>
+#include <DataTypes/DataTypeDateTime64.h>
+#include <Core/Types.h>
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnTuple.h>
+#include <Columns/ColumnString.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_REQUEST_PARAMETER;
+    extern const int NOT_IMPLEMENTED;
+}
+
+PrometheusHTTPProtocolAPI::PrometheusHTTPProtocolAPI(ConstStoragePtr time_series_storage_, const ContextPtr & context_)
+    : WithContext{context_}
+    , time_series_storage(storagePtrToTimeSeries(time_series_storage_))
+    , log(getLogger("PrometheusHTTPProtocolAPI"))
+{
+}
+
+PrometheusHTTPProtocolAPI::~PrometheusHTTPProtocolAPI() = default;
+
+void PrometheusHTTPProtocolAPI::executePromQLQuery(
+    WriteBuffer & response,
+    const Params & params)
+{
+    try
+    {
+        auto query_tree = std::make_unique<PrometheusQueryTree>();
+        query_tree->parse(params.promql_query);
+        if (!query_tree)
+        {
+            writeErrorResponse(response, "bad_data", "Failed to parse PromQL query");
```

Comment:
I'd prefer to just throw an exception here and catch it in `handlingRequestWithContext()`.